src/mono/dbi/cordb-assembly.cpp

```diff
@@ -0,0 +1,395 @@
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbModule - EnableJITDebugging - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT CordbModule::EnableClassLoadCallbacks(BOOL bClassLoadCallbacks)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbModule - EnableClassLoadCallbacks - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT CordbModule::GetFunctionFromToken(mdMethodDef methodDef, ICorDebugFunction** ppFunction)
+{
+    HRESULT hr = S_OK;
+    EX_TRY 
+    {
+        LOG((LF_CORDB, LL_INFO1000000, "CordbModule - GetFunctionFromToken - IMPLEMENTED\n"));
+        MdbgProtBuffer localbuf;
+        m_dbgprot_buffer_init(&localbuf, 128);
+        m_dbgprot_buffer_add_id(&localbuf, m_debuggerId);
+        m_dbgprot_buffer_add_int(&localbuf, methodDef);
+        int cmdId = conn->SendEvent(MDBGPROT_CMD_SET_ASSEMBLY, MDBGPROT_CMD_ASSEMBLY_GET_METHOD_FROM_TOKEN, &localbuf);
+        m_dbgprot_buffer_free(&localbuf);
+
+        ReceivedReplyPacket* received_reply_packet = conn->GetReplyWithError(cmdId);
+        CHECK_ERROR_RETURN_FALSE(received_reply_packet);
+        MdbgProtBuffer* pReply = received_reply_packet->Buffer();
+
+        int            id   = m_dbgprot_decode_id(pReply->p, &pReply->p, pReply->end);
+        CordbFunction* func = NULL;
+        func                = m_pProcess->FindFunction(id);
+        if (func == NULL)
+        {
+            func = new CordbFunction(conn, methodDef, id, this);
+        }
+        func->QueryInterface(IID_ICorDebugFunction, (void**)ppFunction);
+    }
+    EX_CATCH_HRESULT(hr);
+    return hr;
+}
+
+HRESULT CordbModule::GetFunctionFromRVA(CORDB_ADDRESS rva, ICorDebugFunction** ppFunction)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbModule - GetFunctionFromRVA - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT CordbModule::GetClassFromToken(mdTypeDef typeDef, ICorDebugClass** ppClass)
+{
+    CordbClass* pClass = new CordbClass(conn, typeDef, GetDebuggerId());
+    pClass->QueryInterface(IID_ICorDebugClass, (void**)ppClass);
+    return S_OK;
+}
+
+HRESULT
+CordbModule::CreateBreakpoint(ICorDebugModuleBreakpoint** ppBreakpoint)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbModule - CreateBreakpoint - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT CordbModule::GetEditAndContinueSnapshot(ICorDebugEditAndContinueSnapshot** ppEditAndContinueSnapshot)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbModule - GetEditAndContinueSnapshot - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT CordbModule::GetMetaDataInterface(REFIID riid, IUnknown** ppObj)
+{
+    if (m_pRegMeta == NULL)
+    {
+        m_pRegMeta = new RegMeta();
+
+        m_pStgdbRW       = new CLiteWeightStgdbRW();
+        ULONG32 pcchName = 0;
+        GetName(0, &pcchName, NULL);
+
+        WCHAR* full_path;
+        full_path = (WCHAR*)malloc(sizeof(WCHAR) * pcchName);
+        GetName(pcchName, &pcchName, full_path);
+
+        m_pStgdbRW->OpenForRead(full_path, NULL, 0, 0);
```

Comment:
The coreclr implementation sets up some thread-safety option [here](https://github.com/dotnet/runtime/blob/4f3545e2d1466512f53a2cd8b0db71667810c083/src/coreclr/debug/di/process.cpp#L1748) and it doesn't appear this code is doing similar work. Its possible without this option that multi-threaded accesses to the metadata API from VS would fail.

Reply:
Thanks, I have just set the same option in our code.